File: 151-reverse-words-in-a-string/reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        int i=s.length()-1;
        string ans="";
        while(i>=0){
            while(i>=0 && s[i]==' ') i--;
            int j=i;
            while(j>=0 && s[j]!=' '){
                j--;
            }
            
            ans+=s.substr(j+1,i-j);
            i=j;
            ans+=" ";

        }
        while(ans!="" && ans.back()==' ') ans.pop_back();
        
        return ans;
    }
};

File: 151-reverse-words-in-a-string/reverse-words-in-a-string.java
class Solution {
    public String reverseWords(String s) {
        int i = s.length() - 1;
        StringBuilder ans = new StringBuilder();

        while (i >= 0) {
            // skip spaces
            while (i >= 0 && s.charAt(i) == ' ') i--;
            if (i < 0) break;

            int j = i;
            // find start of word
            while (j >= 0 && s.charAt(j) != ' ') j--;

            // append the word
            ans.append(s.substring(j + 1, i + 1));
            ans.append(" ");

            i = j;
        }

        // remove last extra space
        return ans.toString().trim();
    }
}
